### ISSUE TITLE
请问这个模型适合三维点的特征值预测吗

### ISSUE THREAD

**Issue by somethingforheheda** [dizhu-gis/SRGCNN#1]
数据是几千个三维点 有xyz坐标 每个坐标有一个目标值   我把这些点的xyz当特征都传给模型 想要预测目标值 但是感觉模型预测的值大部分都是一样的数比如都是在13附近  导致mse很高  我把模型改了好多 但是都很难降低到想要的效果 不知道该怎么改了

**Comment by dizhu-gis** [comment]
Definitely yes, you may construct a graph with x,y, and z as the coordinates to construct the spatial structure for those points, and then use the graph as the input for the SRGCNN model. 
Are you able to post some of your code here? Otherwise, it can be hard to know where the issue might be for your model.

**Comment by somethingforheheda** [comment]
我先说一下我大致是怎么改的，因为数据每一行是xyz和目标p值。一共几千行，代表几千个空间点。但是我没办法像您那样用KNN把最邻近的20个节点于中心节点连接来构建邻接矩阵。于是我主要是对邻接矩阵的生成进行了修改。首先我用距离反比方法生成adj，即计算每一个点与其他点的距离distance，然后用1/distance得到权重，但是对于距离很近的权重会出现inf值，我于是改成用e^(-distance)来求权重，第二种构建adj的方法就是用两个点的距离如果小于某个值，就设置它们的权重为1。然后我把x,y,z的值作为您代码中的x_tensor，即输入。把目标p值作为y_tensor，即要预测的值。然后在train_loss = step(model, optimizer, x_tensor, train_idx, y_tensor)中我把idx_all改成了train_idx，这是我的训练集的index范围。在global_MAPE=evaluate_MAPE(model, x_tensor, val_idx,y_tensor)中，把idx_all改成val_idx，这是我验证集的index范围。然后在output, target = return_results(model, x_tensor, val_idx, y_tensor)中把idx_all改成val_idx，我还没有设置测试集，只用了训练集和验证集。 关于数据集的归一化，我也采用了很多方法，比如行归一化，列归一化等，归一化范围为(0,1)还有(-1,1)，但是都最后预测的结果MSE都达到了400以上，其他同类方法，MSE都在80左右。最后，我删除了手动生成adj的方法，我用了一个自适应的邻接矩阵，在GWGCN和GCN的中先定义一个矩阵，self.nodevec1 = nn.Parameter(torch.randn(3000, 300).to(device), requires_grad=True).to(device)，其中3000是我的节点个数，然后在forward 第一行添加adj = F.softmax(F.relu(torch.mm(self.nodevec1, self.nodevec1.t())), dim=1)，这样通过矩阵和其转置矩阵相乘，得到一个3000*3000的adj，并且会在迭代过程中不断改变，最后把MSE 400 勉强降低到了 200，但是效果还是没有达到其他同类的方法，比如idw方法。这些基本是我所有修改的思路了，如果您想要具体代码，可以把邮箱发给我，我把整个代码给您，还有一个模拟的数据集。如果有错误，感谢您的指正！

**Comment by somethingforheheda** [comment]
还想问一个问题，您说的and then use the graph as the input for the SRGCNN model.怎么把构建的图作为输入啊？

**Comment by somethingforheheda** [comment]
还想问一个问题，您说的and then use the graph as the input for the SRGCNN model.怎么把构建的图作为输入啊？
构建的图应该就是adj吧？adj已经是一个输入模型的参数了。

**Comment by somethingforheheda** [comment]
代码中的模型输入是用经纬度构建的adj和经纬度对应点的4个标签值作为输入，我这个用xyz距离构建adj，然后在用xyz点作为输入吗（我在实验中就是这么做的）？但是我不懂您说的and then use the graph as the input for the SRGCNN model.怎么把构建的图作为输入啊？构建的图应该就是adj吧？

**Comment by dizhu-gis** [comment]
Yes, adj is the graph, and the graph as a tensor is used in the model's forward function for feature propagation.

**Comment by dizhu-gis** [comment]
> 

Did you try using SRGCNN-GW and SRGCNN to see whether there is a difference in your final accuracy? Maybe more features should be used as the covariates?

**Comment by somethingforheheda** [comment]
我将问下您论文中的adj是怎么构建的啊，就是将每个点与周围最近的20个点进行连接，设置它们的边权重为1吗？

**Comment by somethingforheheda** [comment]
我有一个问题，您代码中对price进行了ln处理，可以看作是归一化操作吗？ 那么最后输出MAPE的时候，应该用反归一化的输出和真实值进行计算吧。于是我在evaluate_MAPE函数中将模型输出的值output和真实值target_tensor进行反归一化(MSE = evaluate_MSE(pow(2.718281828459,output[idx]), pow(2.718281828459,target_tensor[idx])))
函数具体如下，应该是没问题的
def evaluate_MSE(records_real, records_predict):
    if len(records_real) == len(records_predict):
        return sum([(x - y) ** 2 for x, y in zip(records_real, records_predict)]) / len(records_real)
    else:
        return None
但是实验结果发现，在global_MAPE已经接近最小时11.9785%，我计算的MSE的结果还是非常大，您可以在您的实验中测试一下。这样的话，这个模型还是无法预测真实的价格吧。

**Comment by somethingforheheda** [comment]
[15.9419],
[15.9415],
[15.9475],
[15.9456],
[15.9450],
[15.9455],
[15.9418],
[15.9417],
[15.9463],
[15.9418],
[15.9412],
[15.9416],
[16.0355],
[15.9462],
[15.9467],
[15.9398],
[15.9420]],
这是我的数据输入后，模型输出的值，很明显这些值都太集中了。我之前做图卷积预测xyz点的时候也是这样的结果。我换了您的数据，发现模型输出的值也是在一个值附近。您的模型和数据得到的结果也是类似的。
[4.8114],
        [5.4469],
        [4.7250],
        ...,
        [4.7460],
        [5.0063],
        [5.0042]
这是您的数据输入到模型后得到的结果，也是在一个值附近。
感觉这是一个问题吧，您的数据是因为做了一个ln(price)处理，得到的值也很集中，才会使得模型得到的精度高吧。

**Comment by somethingforheheda** [comment]
[15.9455],
[15.9418],
[15.9417],
[15.9463],
[15.9418],
[15.9412],
[15.9416],
[16.0355],
[15.9462],
[15.9467],
[15.9398],
[15.9420]],
这是我的数据输入后，模型输出的值，很明显这些值都太集中了。我之前做图卷积预测xyz点的时候也是这样的结果。我换了您的数据，发现模型输出的值也是在一个值附近。您的模型和数据得到的结果也是类似的。
[4.8114],
[5.4469],
[4.7250],
...,
[4.7460],
[5.0063],
[5.0042]
这是您的数据输入到模型后得到的结果，也是在一个值附近。
感觉这是一个问题吧，您的数据是因为做了一个ln(price)处理，得到的值也很集中，才会使得模型得到的精度高吧。这个貌似是图神经网络的过平滑问题，我不知道怎么解决了

**Comment by northward111** [comment]
> 代码中的模型输入是用经纬度构建的adj和经纬度对应点的4个标签值作为输入，我这个用xyz距离构建adj，然后在用xyz点作为输入吗（我在实验中就是这么做的）？但是我不懂您说的and then use the graph as the input for the SRGCNN model.怎么把构建的图作为输入啊？构建的图应该就是adj吧？

我觉得用点的坐标值作为feature有点奇怪，点的特征应该是决定目标变量的变量，点的坐标只能算是对这一决定过程有一点影响 的变量，决定性变量和影响性变量的区别？或许这个模型不适合仅仅只有坐标值和待预测变量的情况。

**Comment by somethingforheheda** [comment]
> > 代码中的模型输入是用经纬度构建的adj和经纬度对应点的4个标签值作为输入，我这个用xyz距离构建adj，然后在用xyz点作为输入吗（我在实验中就是这么做的）？但是我不懂您说的and then use the graph as the input for the SRGCNN model.怎么把构建的图作为输入啊？构建的图应该就是adj吧？
> 
> 我觉得用点的坐标值作为feature有点奇怪，点的特征应该是决定目标变量的变量，点的坐标只能算是对这一决定过程有一点影响 的变量，决定性变量和影响性变量的区别？或许这个模型不适合仅仅只有坐标值和待预测变量的情况。

我这个数据，就是点的位置决定点的待预测值的（比如2个点距离很近他们的待预测值就应该很接近）。问题是图卷积网络需要一个邻接矩阵作为输入，如果利用距离反比作为权重连接大部分节点，就会导致过度平滑，最后预测的值都差不多（这篇论文感觉也有过度平滑的问题）。如果连接少部分点，又会导致距离信息利用得不够充分。结果就是我最后预测的值基本都一样。
